### PR TITLE
[Toolchain] Revisit ToolchainProvider class

### PR DESCRIPTION
**Pull request by YongseopKim**
This commit revisits ToolchainProver class in src/Toolchain.

### CODE REVIEW

**Comment by YongseopKim** [comment]
The points
- ToolchainProvider can be used as TreeDataProvider as possible
- To do it, let's introduce Node classes(BaseNode <- BackendNode or ToolchainNode)
- BackendNode is an expression for Backend `{}` in ToolchainView
- ToolchainNode is an expression for Toolchain `o` in ToolchainView
- Let's introduce NodeBuilder btw Node and Provider

**Comment by YongseopKim** [comment]
PTAL @jyoungyun

File: src/Tests/Backend/Backend.test.ts [hunk 1]
```diff
@@ -48,9 +50,18 @@ suite('Backend', function() {
...
       // this runs once
       for (const [key, value] of entries) {
-        assert.strictEqual(key, 'Mockup');
+        assert.strictEqual(key, backendName);
         assert.deepStrictEqual(value, backend);
       }
     });
   });
+
+  teardown(function() {
```
Comment (YongseopKim) [inline src/Tests/Backend/Backend.test.ts:59]: This part can be separated pr.

File: src/Tests/Toolchain/ToolchainEnv.test.ts [hunk 2]
```diff
@@ -21,8 +21,9 @@ import {ToolchainInfo, Toolchains} from '../../Backend/Toolchain';
 import {DebianToolchain} from '../../Backend/ToolchainImpl/DebianToolchain';
 import {ToolchainEnv} from '../../Toolchain/ToolchainEnv';
 
-const mocCompilerType: string = 'test';
 
+// TODO: Move this to MockCompiler.ts
```
Comment (YongseopKim) [inline src/Tests/Toolchain/ToolchainEnv.test.ts:25]: ditto

File: src/Toolchain/ToolchainProvider.ts [hunk 3]
```diff
@@ -17,88 +17,105 @@
 import * as vscode from 'vscode';
 
 import {Toolchain} from '../Backend/Toolchain';
-import {DebianToolchain} from '../Backend/ToolchainImpl/DebianToolchain';
-import {Job, JobCallback} from '../Project/Job';
+import {Job} from '../Project/Job';
```
Comment (YongseopKim) [inline src/Toolchain/ToolchainProvider.ts:20]: I think it's better to read this file as one file not diff.

File: src/Tests/Backend/Backend.test.ts (hunk 1 above)
Comment (YongseopKim) [reply src/Tests/Backend/Backend.test.ts:59]: https://github.com/Samsung/ONE-vscode/pull/956

File: src/Tests/Toolchain/ToolchainEnv.test.ts (hunk 2 above)
Comment (YongseopKim) [reply src/Tests/Toolchain/ToolchainEnv.test.ts:25]: https://github.com/Samsung/ONE-vscode/pull/957

**Review by jyoungyun — DISMISSED** [review]
LGTM

**Comment by YongseopKim** [comment]
PTAL @Samsung/one-vscode

**Comment by YongseopKim** [comment]
Rebased. PTAL @jyoungyun and @Samsung/one-vscode

**Review by jyoungyun — DISMISSED** [review]
LGTM

**Comment by YongseopKim** [comment]
Rebased again... PTAL @jyoungyun and @Samsung/one-vscode

**Review by jyoungyun — APPROVED** [review]
LGTM

Please review this PR. /cc @Samsung/one-vscode

**Review by llFreetimell — APPROVED** [review]
LGTM

**Pull request merged**